doc/internals.rst

```diff
@@ -231,3 +231,410 @@ re-open it directly with Zarr:
+grouped in three types of indexes
+:py:class:`~xarray.core.indexing.BasicIndexer`,
+:py:class:`~xarray.core.indexing.OuterIndexer` and
+:py:class:`~xarray.core.indexing.VectorizedIndexer`.
+This implies that the implementation of the method ``__getitem__`` can be tricky.
+In oder to simplify this task, Xarray provides a helper function,
+:py:func:`~xarray.core.indexing.explicit_indexing_adapter`, that transforms
+all the input  ``indexer`` types (`basic`, `outer`, `vectorized`) in a tuple
+which is interpreted correctly by your backend.
+
+This is an example ``BackendArray`` subclass implementation:
+
+.. code-block:: python
+
+    class YourBackendArray(BackendArray):
+        def __init__(
+            self,
+            # other backend specific keyword arguments
+        ):
+            self.shape = ...
+            self.dtype = ...
+            self.lock = ...
+
+        def __getitem__(self, key):
+            return indexing.explicit_indexing_adapter(
+                key,
+                self.shape,
+                indexing.IndexingSupport.BASIC,
+                self._raw_indexing_method,
+            )
+
+        def _raw_indexing_method(self, key):
+            # thread safe method that access to data on disk
+            with self.lock:
+                ...
+                return item
+
+Note that ``BackendArray.__getitem__`` must be thread safe to support
+multi-thread processing.
+
+The :py:func:`~xarray.core.indexing.explicit_indexing_adapter` method takes in
+input the ``key``, the array ``shape`` and the following parameters:
+
+- ``indexing_support``: the type of index supported by ``raw_indexing_method``
+- ``raw_indexing_method``: a method that shall take in input a key in the form
+  of a tuple and return an indexed :py:class:`numpy.ndarray`.
+
+For more details see
+:py:class:`~xarray.core.indexing.IndexingSupport` and :ref:`RST indexing`.
+
+In order to support `Dask <http://dask.pydata.org/>`__ distributed and
+:py:mod:`multiprocessing`, ``BackendArray`` subclass should be serializable
+either with :ref:`io.pickle` or
+`cloudpickle <https://github.com/cloudpipe/cloudpickle>`__.
+That implies that all the reference to open files should be dropped. For
+opening files, we therefore suggest to use the helper class provided by Xarray
+:py:class:`~xarray.backends.CachingFileManager`.
+
+.. _RST indexing:
+
+Indexing Examples
+^^^^^^^^^^^^^^^^^
+**BASIC**
+
+In the ``BASIC`` indexing support, numbers and slices are supported.
+The behaviour is the same as `NumPy <https://numpy.org/>`__ .
+
+Example:
+
+.. code-block:: python
+
+    # () shall return the full array
+    >>> backend_array._raw_indexing_method(())
+    array([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
+
+    # shall support integers
+    >> backend_array._raw_indexing_method(1, 1)
+    5
+
+    # shall support slices
+    >>> backend_array._raw_indexing_method(slice(0, 3), slice(2, 4))
+    array([[2, 3], [6, 7], [10, 11]])
```

Comment:
while `sphinx` supports this, `blackdoc` was not written with something like this in mind (and adding this will probably take some time). Since we normally use `ipython` instead, how about converting this to verbatim `ipython`? For example, this would become:
```suggestion
.. ipython::
    :verbatim:
    
    In [1]: # () shall return the full array
       ...: backend_array._raw_indexing_method(())
    Out[1]: array([[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]])
    
    In [2]: # shall support integers
       ...: backend_array._raw_indexing_method(1, 1)
    Out[2]: 5
    
    In [3]: # shall support slices
       ...: backend_array._raw_indexing_method(slice(0, 3), slice(2, 4))
    Out[3]: array([[2, 3], [6, 7], [10, 11]])
```

Reply:
thank you very much for your suggestions!
However, it is not worth reviewing the new documentation now, sincee the interface for the lazy loading part will probably be changed.

Reply:
sure. This is more a hint of how to get the linting CI to pass.